### PR TITLE
BUG: input t_r had no effect in clean_img and signal!

### PR DESCRIPTION
**Pull request by raamana**

### CODE REVIEW

**Comment by raamana** [comment]
Not sure why that function had to be so dense - saving memory or improving speed? 

Perhaps `clean_img` routine could wear its name better and be a bit more clean and lean :), like:

```python
    imgs_ = check_niimg_4d(imgs)
    voxel_by_time = imgs_.get_data().reshape(-1, imgs_.shape[-1]).T
    cleaned_data = signal.clean(voxel_by_time, sessions=sessions,
        detrend=detrend, standardize=standardize, confounds=confounds,
        low_pass=low_pass, high_pass=high_pass, t_r=t_r,
        ensure_finite=ensure_finite)
    cleaned_image = new_img_like(imgs, cleaned_data.T.reshape(imgs_.shape), copy_header=True)
    return cleaned_image
```
which is still dense, but one can make the gist of the function better.

**Comment by bthirion** [comment]
This was a bad bug indeed ! Thx.

**Comment by GaelVaroquaux** [comment]
OMG, that's bad. Thanks heap.

We need a test. Can you add one please?

**Comment by KamalakerDadi** [comment]
Could you please add some tests. Thanks a lot for spotting.

**Comment by raamana** [comment]
No problem. I will be happy to. I assume I've a week or two for this? :)

**Comment by raamana** [comment]
Just caught something else in f7fac80 - this might not be a serious bug, as the code never reaches that point when using `nilearn.signal.clean` or `nilearn.image.clean_img`, but is a bug if someone tries to use `nilearn.signal.butterworth` directly. Perhaps worth considering making it private `nilearn.signal._butterworth` private, if its not meant for the general user?

File: nilearn/signal.py [hunk 1]
```diff
@@ -168,22 +166,22 @@ def _detrend(signals, inplace=False, type="linear", n_batches=10):
...
+        warnings.warn(
+            'The frequency specified for the %s pass filter is '
+            'too high to be handled by a digital filter (superior to '
+            'nyquist frequency). It has been lowered to %.2f (nyquist '
+            'frequency).' % (btype, wn))
     return wn
 
 
-def butterworth(signals, sampling_rate, low_pass=None, high_pass=None,
+def butterworth(signals_in, sampling_rate, low_pass=None, high_pass=None,
```
Comment (GaelVaroquaux) [inline nilearn/signal.py]: Why did you change this? We'd rather keep the name "signals".

**Comment by GaelVaroquaux** [comment]
Nice catch with regards to signal vs signals! Once again, this warrants a test. The codebase is reveal how much the old saying is true: "if it ain't tested, it's broken".

**Comment by raamana** [comment]
Yes and yes.. I’m working towards the test, trying to understand what’s going on before I write something quickly.. 

From what I can see now, I just need to check if the result

**Comment by raamana** [comment]
Yes and yes.. I’m working towards the test, trying to understand what’s going on before I write something quickly..

From what I can see now, I just need to write a test to ensure result for different TRs are not the same, as other smaller components are already tested.

**Comment by GaelVaroquaux** [comment]
> From what I can see now, I just need to write a test to ensure result for
> different TRs are not the same, as other smaller components are already tested.

That seems right to me.

**Comment by raamana** [comment]
Great. What are the typical ranges for low- and high-pass cutoff frequencies used in fMRI/nilearn?

File: nilearn/signal.py (hunk 1 above)
Comment (raamana) [reply nilearn/signal.py]: I originally did it to disambiguate it from the signal import from scipy, but now that I renamed that also, this is not necessary :)

**Comment by raamana** [comment]
This [link](https://en.wikibooks.org/wiki/Neuroimaging_Data_Processing/Temporal_Filtering) seems to suggest there are no rules per se, but the frequency ranges are tied to sampling rate (aka TR).

Given I am testing it on noise, perhaps this is not important.

File: nilearn/tests/test_signal.py [hunk 2]
```diff
@@ -274,6 +287,35 @@ def test_clean_detrending():
     assert_false(abs(x_undetrended - signals).max() < 0.06)
 
 
+def test_clean_TR():
```
Comment (KamalakerDadi) [inline nilearn/tests/test_signal.py]: I would keep small letters.

File: nilearn/signal.py [hunk 3]
```diff
@@ -168,23 +166,31 @@ def _detrend(signals, inplace=False, type="linear", n_batches=10):
 def _check_wn(btype, freq, nyq):
     wn = freq / float(nyq)
     if wn >= 1.:
-        warnings.warn(
```
Comment (KamalakerDadi) [inline nilearn/signal.py]: Do you have any strong reason to remove from here and add below ?

Comment (raamana) [reply nilearn/signal.py]: I need the value for wn to print it - I was fixing a bug in the warning message.

**Comment by raamana** [comment]
Admins, I am yet to get a sense of the "nilearn" style - if you see something minor changed, please feel free to edit it.

**Comment by KamalakerDadi** [comment]
Ok. Please let us know when to review. I am happy to review.

**Comment by raamana** [comment]
Thanks man.

Everyone, the current test is failing for *few random pairs* of TRs on the CI servers, although it passed locally on my laptop so far. Assuming this is not specific to py2.7 or CI, we might need to a devise "a more methodical approach" to establishing the expectations (for what values of TRs are the results expected to be different etc), to reduce surprises.

File: nilearn/tests/test_signal.py [hunk 4]
```diff
@@ -133,6 +143,8 @@ def test_butterworth():
                                 low_pass=low_pass, high_pass=high_pass,
                                 copy=True)
     np.testing.assert_almost_equal(data, data_original)
+    np.testing.assert_(id(out1)!=id(data_original))
```
Comment (GaelVaroquaux) [inline nilearn/tests/test_signal.py]: Python coding style (known as "PEP8") says that you should have a space before and after the "!=" operator.

**Comment by raamana** [comment]
I feel we need to replicate this test in `nilearn/image/test_image.py` to ensure `clean_img` calls `nilearn.signal.clean` correctly (pattern uncovering the original bug). Any better ways to achieve it without duplicating the testing code?

File: nilearn/tests/test_signal.py [hunk 5]
```diff
@@ -274,6 +287,40 @@ def test_clean_detrending():
...
+def test_clean_t_r():
+    """Different TRs must produce different results after filtering"""
+
+    # n_features  Must be higher than 500
+    for n_samples, n_features in zip(( 34,  42, 100),
+                                     (501, 647, 786)):
+        x_orig = generate_signals_plus_trends(n_features=n_features, n_samples=n_samples)
+
+        random_tr_list1 = np.round(np.random.rand(10)*10, decimals=2)
+        random_tr_list2 = np.round(np.random.rand(10)*10, decimals=2)
```
Comment (GaelVaroquaux) [inline nilearn/tests/test_signal.py]: Here you have an unseeded random-number generator. This will create instabilities in the tests that may lead to failures that are hard to reproduce.

The way that we use to avoid this problem is that we create a local random number generator, using "rng = np.random.RandomState(0)". "rng" is then a random number generator from which we can drawn samples using "rng.rand".

This could help stabilize the tests.

**Comment by GaelVaroquaux** [comment]
> Any better ways to achieve it without duplicating the testing code?

I don't think that you need a test as thorough: you are testing many
variants of parameters. I would do something simpler.

File: nilearn/tests/test_signal.py (hunk 5 above)
Comment (raamana) [reply nilearn/tests/test_signal.py]: True. But then we are reducing the test parameter space - sampling only the same few fixed values. I'm just trying to catch all edge cases now than later. cc @bthirion.

**Comment by bthirion** [comment]
"What are the typical ranges for low- and high-pass cutoff frequencies used in fMRI/nilearn?"
low: .1 hz, high: .01hz.

File: nilearn/signal.py (hunk 3 above)
Comment (KamalakerDadi) [reply nilearn/signal.py]: What is the bug in the warning message ?

File: nilearn/signal.py [hunk 6]
```diff
@@ -252,10 +258,10 @@ def butterworth(signals, sampling_rate, low_pass=None, high_pass=None,
     else:
         critical_freq = critical_freq[0]
 
-    b, a = signal.butter(order, critical_freq, btype=btype)
+    b, a = sp_signal.butter(order, critical_freq, btype=btype, output='ba')
```
Comment (KamalakerDadi) [inline nilearn/signal.py]: Importing as sp_signal may not require I suppose ?

Comment (raamana) [reply nilearn/signal.py]: helps clarifying its different from input parameter signals. I was gonna call it scipy_signal, but then it seemed too long :)

File: nilearn/signal.py (hunk 3 above)
Comment (raamana) [reply nilearn/signal.py]: The warning message does not reflect action taken inside.

**Comment by raamana** [comment]
thanks Bertrand for giving me the typical values - could you provide me with ranges for them?

**Comment by bthirion** [comment]
Not sure I understand your request: you don't want to check the values entered by users, otherwise the code will become a complete mess.
Just fYI the values would be [.0, .02] for high pass and [.05, .5] for low pass.
Note that you can do your analysis without band-pass filtering: it may just be a bit more noisy.

**Comment by raamana** [comment]
Thanks, I don't plan to use them to check the user's  input. Just to use them as part of the testing process: 

https://github.com/raamana/nilearn/blob/0cc967c53279b87cc7295ebeb0cd6b66d15e3fdd/nilearn/tests/test_signal.py#L301-304

**Comment by raamana** [comment]
NP Kamalaker. TBH, I don't understand this conflict well - what is its problem if I rename an import? I think importing `scipy.signal` as `signal` is dangerous, given the abundant use of the word `signals` as variable name. I'd advocate for `sp_signal` or anything else suitable that is much less ambiguous.

**Comment by KamalakerDadi** [comment]
>TBH, I don't understand this conflict well - what is its problem if I rename an import?

You got conflicts because @GaelVaroquaux had directly modified the line of imports in master. And you are also trying to modify the line of imports. That's why a conflict. Could you please resolve it by rebasing on top of master ? Thanks

>I think importing scipy.signal as signal is dangerous, given the abundant use of the word signals as variable name. I'd advocate for sp_signal or anything else suitable that is much less ambiguous.

Agreed.

**Comment by bthirion** [comment]
Thx for the contribution !

File: nilearn/signal.py (hunk 1 above)
Comment (KamalakerDadi) [reply nilearn/signal.py]: This is addressed in PR #1656

File: nilearn/tests/test_signal.py [hunk 7]
```diff
@@ -274,6 +287,40 @@ def test_clean_detrending():
     assert_false(abs(x_undetrended - signals).max() < 0.06)
 
 
+def test_clean_t_r():
+    """Different TRs must produce different results after filtering"""
+
+    # n_features  Must be higher than 500
+    for n_samples, n_features in zip((34, 42, 100),
```
[the remark this answers is not in the document]
Comment (KamalakerDadi) [reply nilearn/tests/test_signal.py]: Addressed in PR #1656 .  Reduced the number of iterations.

**Comment by GaelVaroquaux** [comment]
Closing as completed in #1656.

Thanks @raamana !

**Comment by raamana** [comment]
Thank you KD for helping me clean it up. Very glad to contribute a bit to nilearn. Looking fwd to much more.

**Comment by GaelVaroquaux** [comment]
We're very happy that you did. This was a nasty bug. Thank you!

**Comment by raamana** [comment]
NP, thank you for merging.
